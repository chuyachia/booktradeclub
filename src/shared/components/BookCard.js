// records books own only in books model and aggregate to show all books own by a user in profile

import {connect} from "react-redux";
import React from "react";
import {viewBook} from "../actions/profileAction";

class BookCard extends React.Component{
    constructor(props){
        super(props);
    }
    viewBooks(){
        this.props.dispatch(viewBook(this.props.info,this.props.modaluse));
    }
    render(){
        return(
            <div class="card bookdisplay" onClick={this.viewBooks.bind(this)}>
                <img src={this.props.info.imageUrl} alt={this.props.info.title} class="img-thumbnail"/>
                  <div class="middle">
                    <i class="fas fa-search fa-lg"></i>
                </div>
            </div>
            );
    }
}

export default connect()(BookCard);