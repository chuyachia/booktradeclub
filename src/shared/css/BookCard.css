.card {
  position: relative;
  height: 150px;
  width: 100px;
  overflow: hidden;
  display:inline-block;
  cursor:pointer;
  margin:5px;
}


.card img {
  position: absolute;
  left: 50%;
  top: 50%;
  height: 100%;
  width: auto;
  transition: .5s ease;
  backface-visibility: hidden;
  opacity:1;
  -webkit-transform: translate(-50%,-50%);
      -ms-transform: translate(-50%,-50%);
          transform: translate(-50%,-50%);
}

.center {
  transition: .5s ease;
  opacity: 0;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  text-align: center;
}

.card:hover img {
  opacity: 0.3;
}

.card:hover .center {
  opacity: 1;
}
