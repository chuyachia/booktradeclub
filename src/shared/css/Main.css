:global(.container) {
    padding-top:80px;
}

:global(.nav-item){
    cursor:pointer
}

:global(img){
  margin-right:5px;
  float:left;
}

:global(.footer) {
  position:fixed;
  bottom:0;
  right:0;
  text-align: right;
  padding:15px
}
:global(.footer) span {
  position:relative;
  display:block
}

:global(.fas), :global(.fab){
  margin-top:5px
}

.wrap{
  margin: 0 auto;
  position: absolute;
  top: 50%;
  left: 50%;
}

