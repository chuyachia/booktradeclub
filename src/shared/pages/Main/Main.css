.dropdown{
    float:right;
}
.dropdown-menu-right {
    right: 0;
    left: auto;
}