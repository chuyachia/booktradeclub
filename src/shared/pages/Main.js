import BookCard from "../components/BookCard";
import { connect } from "react-redux";
import filter from "lodash.filter";
import Footer from "../components/Footer";
import {getAllBooks} from "../actions/bookAction";
import Modal from "../components/Modal";
import NavBar from "../components/NavBar";
import React from "react";
import { SyncLoader } from 'react-spinners';
import styles from "../css/Main.css";




class Main extends React.Component {
    constructor(){
        super();
        this.state={term:''};
        this.addSearchTerm = this.addSearchTerm.bind(this);
    }
    addSearchTerm(event){
        this.setState({term: event.target.value});
    }
    componentDidMount(){
        this.props.getAllBooks();
    }
    render(){
        var books = filter(this.props.books,(book)=>book.title.toLowerCase().indexOf(this.state.term)>-1);
        return (
         <div class="container">
            <NavBar/>
            <h1 class="display-4">Books for exchange</h1>
            <input class="searchbook" onChange={this.addSearchTerm} type="search" class="form-control" autoFocus
            placeholder="Enter a book name"/>
            <div class={`${styles.wrap}`}>
                <SyncLoader
                  color={'#424242'} 
                  loading={this.props.loading} 
                />
            </div>
            {this.props.error&&<p>Oops, something went wrong. Please come back later.</p>}
            {books.map(book=> (
                <BookCard key={book._id} info={book} modaluse="addrequest"/>
                ))}
            <Modal/>
            <Footer/>
        </div>);
    }
}

var dispatchMap = {getAllBooks};

var propsMap = (store)=>{
    return {
        loading:store.allbooks.loading,
        books : store.allbooks.books,
        error : store.allbooks.error,
        username : store.userinfo.username
    };
};

export default connect(propsMap,dispatchMap)(Main);
