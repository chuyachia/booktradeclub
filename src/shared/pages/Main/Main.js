// Add books in Search cause Main render loop

import BookCard from "../../components/BookCard";
import { connect } from "react-redux";
import filter from "lodash.filter";
import {getAllBooks} from "../../actions/mainAction";
import { Link } from 'react-router-dom'
import Modal from "../../components/Modal"
import React from "react";

class Main extends React.Component {
    constructor(){
        super();
        this.state={term:''};
    }
    addSearchTerm(event){
        this.setState({term: event.target.value});
    }
    componentWillMount(){
        if (this.props.books.length==0){
            this.props.dispatch(getAllBooks());
        }
    }
    render(){
        console.log(this.props.books);
        var books = filter(this.props.books,(book)=>book.title.toLowerCase().indexOf(this.state.term)>-1);
        return (
        <div>
        {this.props.username&&<div>Hi, {this.props.username}</div>}
        {this.props.username?<a href="/logout">Logout</a>:<Link to="/connect">Login</Link>}
        {this.props.username&&<Link to="/profile">Profile</Link>}
        <input onChange={this.addSearchTerm.bind(this)} type="search" class="form-control" placeholder="Enter a book name"/>
        {books.map((book,i)=> (
            <BookCard key={i} info={book}/>
            ))}
        <Modal btnuse="addrequest"/>
        </div>);
    }
}


var propsMap = (store)=>{
    return {
        books : store.queries.books,
        error : store.queries.error,
        username : store.userinfo.username
    };
};

export default connect(propsMap)(Main);
