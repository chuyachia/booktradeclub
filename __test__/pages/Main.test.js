import '../enzymeSetup';
import React from 'react';
import { mount } from 'enzyme';
import EnzymeToJson from 'enzyme-to-json';
import { Provider } from 'react-redux';
import reducer from "../../src/shared/reducers";
import { BrowserRouter as Router } from 'react-router-dom';
import thunk from "redux-thunk";
import { applyMiddleware, createStore } from 'redux';
import Main from '../../src/shared/pages/Main';
import BookCard from '../../src/shared/components/BookCard';

const flushPromises = () => new Promise(resolve => setImmediate(resolve));

describe('Main page test',()=>{
    let store;
    let component;
    beforeEach(async ()=>{
        store=  applyMiddleware(thunk)(createStore)(reducer);
        component = mount(<Router><Provider store={store}><Main/></Provider></Router>);
        await flushPromises();
        component.update();
    });    
    it('Main page render n BookCards',()=>{
        expect(component.find(BookCard)).toHaveLength(2);
    }); 
    
    it('Main page snapshot',()=>{
        expect(EnzymeToJson(component)).toMatchSnapshot();
    });  
})