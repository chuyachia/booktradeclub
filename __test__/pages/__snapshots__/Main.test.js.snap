// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main page test Main page snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "blank",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <Connect(Main)>
        <Main
          books={
            Array [
              Object {
                "__v": 0,
                "_id": "5b8053fbdadc6d004e76eb4a",
                "authors": Array [
                  "Jean-Claude Marsan",
                ],
                "bookId": "fVEeYOKjvfcC",
                "categories": Array [
                  "Architecture",
                ],
                "description": "Montreal in Evolution presents the rich and complex history of Montreal's architectural and environmental development from the first fort of Ville-Marie to the skyscrapers of today. It also examines the forces which shaped the city during the past three hundred and fifty years.",
                "imageUrl": "http://books.google.com/books/content?id=fVEeYOKjvfcC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
                "ownBy": Array [
                  "test1",
                ],
                "publishedDate": "1990-09-01",
                "publisher": "McGill-Queen's Press - MQUP",
                "title": "Montreal in Evolution",
              },
              Object {
                "__v": 0,
                "_id": "5b996ac617e191132e118ed9",
                "authors": Array [
                  "Jenny E. Tesar",
                ],
                "bookId": "9F20v4CzhDUC",
                "categories": Array [
                  "Juvenile Nonfiction",
                ],
                "description": "Describes the physical characteristics, lifestyle, habitat, reproductive process, and methods of survival of this kangaroo which is native to Australia.",
                "imageUrl": "http://books.google.com/books/content?id=9F20v4CzhDUC&printsec=frontcover&img=1&zoom=1&source=gbs_api",
                "ownBy": Array [
                  "admin",
                ],
                "publishedDate": "1997",
                "publisher": "Blackbirch PressInc",
                "title": "What on Earth is a Quokka?",
              },
            ]
          }
          error={false}
          getAllBooks={[Function]}
          loading={false}
          username={null}
        >
          <div
            className="container"
          >
            <NavBar>
              <nav
                className="navbar navbar-dark bg-dark fixed-top"
              >
                <Link
                  className="navbar-brand"
                  replace={false}
                  to="/"
                >
                  <a
                    className="navbar-brand"
                    href="/"
                    onClick={[Function]}
                  >
                    Book Trading Club
                  </a>
                </Link>
                <Connect(DropDownBtn)>
                  <DropDownBtn
                    dispatch={[Function]}
                    username={null}
                  >
                    <div
                      className="show floatRight"
                    >
                      <a
                        className="btn bg-dark text-light"
                        id="dropdownMenuLink"
                        onClick={[Function]}
                        role="button"
                      >
                        <i
                          className="fas fa-bars"
                        />
                      </a>
                      <div
                        aria-labelledby="dropdownMenuLink"
                        className="dropdown-menu  right"
                        onClick={[Function]}
                      >
                        <Link
                          className="dropdown-item"
                          replace={false}
                          to="/"
                        >
                          <a
                            className="dropdown-item"
                            href="/"
                            onClick={[Function]}
                          >
                            Home
                          </a>
                        </Link>
                        <Link
                          className="dropdown-item"
                          replace={false}
                          to="/connect"
                        >
                          <a
                            className="dropdown-item"
                            href="/connect"
                            onClick={[Function]}
                          >
                            Login
                          </a>
                        </Link>
                        <div
                          className="dropdown-divider"
                        />
                        <Link
                          className="dropdown-item"
                          replace={false}
                          to="/about"
                        >
                          <a
                            className="dropdown-item"
                            href="/about"
                            onClick={[Function]}
                          >
                            About
                          </a>
                        </Link>
                      </div>
                    </div>
                  </DropDownBtn>
                </Connect(DropDownBtn)>
              </nav>
            </NavBar>
            <h1
              className="display-4"
            >
              Books for exchange
            </h1>
            <input
              autoFocus={true}
              className="form-control"
              onChange={[Function]}
              placeholder="Enter a book name"
              type="search"
            />
            <div
              className="wrap"
            >
              <onlyUpdateForKeys(Loader)
                color="#424242"
                loading={false}
                margin="2px"
                size={15}
              >
                <Loader
                  color="#424242"
                  loading={false}
                  margin="2px"
                  size={15}
                />
              </onlyUpdateForKeys(Loader)>
            </div>
            <Connect(BookCard)
              info={
                Object {
                  "__v": 0,
                  "_id": "5b8053fbdadc6d004e76eb4a",
                  "authors": Array [
                    "Jean-Claude Marsan",
                  ],
                  "bookId": "fVEeYOKjvfcC",
                  "categories": Array [
                    "Architecture",
                  ],
                  "description": "Montreal in Evolution presents the rich and complex history of Montreal's architectural and environmental development from the first fort of Ville-Marie to the skyscrapers of today. It also examines the forces which shaped the city during the past three hundred and fifty years.",
                  "imageUrl": "http://books.google.com/books/content?id=fVEeYOKjvfcC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
                  "ownBy": Array [
                    "test1",
                  ],
                  "publishedDate": "1990-09-01",
                  "publisher": "McGill-Queen's Press - MQUP",
                  "title": "Montreal in Evolution",
                }
              }
              key="5b8053fbdadc6d004e76eb4a"
              modaluse="addrequest"
            >
              <BookCard
                handleClick={[Function]}
                info={
                  Object {
                    "__v": 0,
                    "_id": "5b8053fbdadc6d004e76eb4a",
                    "authors": Array [
                      "Jean-Claude Marsan",
                    ],
                    "bookId": "fVEeYOKjvfcC",
                    "categories": Array [
                      "Architecture",
                    ],
                    "description": "Montreal in Evolution presents the rich and complex history of Montreal's architectural and environmental development from the first fort of Ville-Marie to the skyscrapers of today. It also examines the forces which shaped the city during the past three hundred and fifty years.",
                    "imageUrl": "http://books.google.com/books/content?id=fVEeYOKjvfcC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
                    "ownBy": Array [
                      "test1",
                    ],
                    "publishedDate": "1990-09-01",
                    "publisher": "McGill-Queen's Press - MQUP",
                    "title": "Montreal in Evolution",
                  }
                }
                modaluse="addrequest"
              >
                <div
                  className="card card"
                  onClick={[Function]}
                >
                  <img
                    alt="Montreal in Evolution"
                    className="img-thumbnail"
                    src="http://books.google.com/books/content?id=fVEeYOKjvfcC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
                  />
                  <div
                    className="center"
                  >
                    <i
                      className="fas fa-search fa-lg"
                    />
                  </div>
                </div>
              </BookCard>
            </Connect(BookCard)>
            <Connect(BookCard)
              info={
                Object {
                  "__v": 0,
                  "_id": "5b996ac617e191132e118ed9",
                  "authors": Array [
                    "Jenny E. Tesar",
                  ],
                  "bookId": "9F20v4CzhDUC",
                  "categories": Array [
                    "Juvenile Nonfiction",
                  ],
                  "description": "Describes the physical characteristics, lifestyle, habitat, reproductive process, and methods of survival of this kangaroo which is native to Australia.",
                  "imageUrl": "http://books.google.com/books/content?id=9F20v4CzhDUC&printsec=frontcover&img=1&zoom=1&source=gbs_api",
                  "ownBy": Array [
                    "admin",
                  ],
                  "publishedDate": "1997",
                  "publisher": "Blackbirch PressInc",
                  "title": "What on Earth is a Quokka?",
                }
              }
              key="5b996ac617e191132e118ed9"
              modaluse="addrequest"
            >
              <BookCard
                handleClick={[Function]}
                info={
                  Object {
                    "__v": 0,
                    "_id": "5b996ac617e191132e118ed9",
                    "authors": Array [
                      "Jenny E. Tesar",
                    ],
                    "bookId": "9F20v4CzhDUC",
                    "categories": Array [
                      "Juvenile Nonfiction",
                    ],
                    "description": "Describes the physical characteristics, lifestyle, habitat, reproductive process, and methods of survival of this kangaroo which is native to Australia.",
                    "imageUrl": "http://books.google.com/books/content?id=9F20v4CzhDUC&printsec=frontcover&img=1&zoom=1&source=gbs_api",
                    "ownBy": Array [
                      "admin",
                    ],
                    "publishedDate": "1997",
                    "publisher": "Blackbirch PressInc",
                    "title": "What on Earth is a Quokka?",
                  }
                }
                modaluse="addrequest"
              >
                <div
                  className="card card"
                  onClick={[Function]}
                >
                  <img
                    alt="What on Earth is a Quokka?"
                    className="img-thumbnail"
                    src="http://books.google.com/books/content?id=9F20v4CzhDUC&printsec=frontcover&img=1&zoom=1&source=gbs_api"
                  />
                  <div
                    className="center"
                  >
                    <i
                      className="fas fa-search fa-lg"
                    />
                  </div>
                </div>
              </BookCard>
            </Connect(BookCard)>
            <Connect(Modal)>
              <Modal
                addBook={[Function]}
                addExchange={[Function]}
                addRequest={[Function]}
                btnuse={null}
                closeBook={[Function]}
                email={null}
                getUsersLocation={[Function]}
                info={
                  Object {
                    "authors": Array [],
                    "bookId": "",
                    "categories": Array [],
                    "description": "",
                    "imageUrl": "",
                    "ownBy": Array [],
                    "publishedDate": "",
                    "publisher": "",
                    "title": "",
                  }
                }
                open={false}
                outrequests={Array []}
                ownedbooks={Array []}
                ownerslocation={Array []}
                removeBook={[Function]}
                tologin={false}
                username={null}
              >
                <div>
                  <Connect(Alert)>
                    <Alert
                      closeAlert={[Function]}
                      open={false}
                    >
                      <div />
                    </Alert>
                  </Connect(Alert)>
                  <div
                    className="overlay "
                  />
                  <div
                    className="modal "
                    style={
                      Object {
                        "display": "none",
                        "overflowY": "scroll",
                      }
                    }
                  >
                    <div
                      className="modal-dialog"
                    >
                      <div
                        className="modal-content"
                      >
                        <div
                          className="modal-header"
                        >
                          <h5
                            className="modal-title"
                          />
                          <button
                            className="close"
                            onClick={[Function]}
                            type="button"
                          >
                            <span>
                              ×
                            </span>
                          </button>
                        </div>
                        <div
                          className="modal-body"
                        >
                          <dl>
                            <dt>
                              <img
                                alt=""
                                className="img-thumbnail"
                                src=""
                              />
                            </dt>
                            <div>
                              <dt>
                                Author
                              </dt>
                              <dd />
                            </div>
                            <dt>
                              Publisher
                            </dt>
                            <dd />
                            <dt>
                              Published
                            </dt>
                            <dd />
                            <dd />
                          </dl>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </Modal>
            </Connect(Modal)>
            <Footer>
              <div
                className="footer"
              >
                <span>
                  <a
                    href="https://github.com/chuyachia/"
                    target="_blank"
                  >
                    <i
                      className="fab fa-github fa-2x text-dark"
                    />
                  </a>
                </span>
                <span>
                  <a
                    href="mailto:chuyachia@gmail.com"
                  >
                    <i
                      className="fas fa-at fa-2x text-dark"
                    />
                  </a>
                </span>
              </div>
            </Footer>
          </div>
        </Main>
      </Connect(Main)>
    </Provider>
  </Router>
</BrowserRouter>
`;
